nebula3/data/ResultSet.py

```diff
@@ -192,6 +192,187 @@ def rows(self):
+                    for k, v in props_raw.items()
+                }
+
+                if "id" not in props:
+                    props["id"] = node_id
+
+                if node_id not in nodes_dict:
+                    nodes_dict[node_id] = {
+                        "id": node_id,
+                        "labels": tags,
+                        "props": props,
+                    }
+                else:
+                    nodes_dict[node_id]["labels"] = list(
+                        set(nodes_dict[node_id]["labels"] + tags)
+                    )
+                    nodes_dict[node_id]["props"].update(props)
+
+            elif isinstance(item, Relationship):
+                src_id = str(item.start_vertex_id().cast())
+                dst_id = str(item.end_vertex_id().cast())
+                rank = item.ranking()
+                edge_name = item.edge_name()
+                props_raw = item.properties()
+                props = {
+                    k: str(v.cast()) if hasattr(v, "cast") else str(v)
+                    for k, v in props_raw.items()
+                }
+                if str((src_id, dst_id, rank, edge_name)) not in edges_dict:
+                    edges_dict[str((src_id, dst_id, rank, edge_name))] = {
+                        "src": src_id,
+                        "dst": dst_id,
+                        "name": edge_name,
+                        "rank": rank,
+                        "props": props,
+                    }
+                else:
+                    edges_dict[(src_id, dst_id, rank, edge_name)]["props"].update(props)
+
+            elif isinstance(item, PathWrapper):
+                for node in item.nodes():
+                    add_to_nodes_or_edges(nodes_dict, edges_dict, node)
+                for edge in item.relationships():
+                    add_to_nodes_or_edges(nodes_dict, edges_dict, edge)
+
+            elif isinstance(item, list):
+                for it in item:
+                    add_to_nodes_or_edges(nodes_dict, edges_dict, it)
+
+        nodes_dict = dict()
+        edges_dict = dict()
+
+        columns = self.keys()
+        for col_num in range(self.col_size()):
+            col_name = columns[col_num]
+            col_list = self.column_values(col_name)
+            add_to_nodes_or_edges(nodes_dict, edges_dict, [x.cast() for x in col_list])
+        nodes = list(nodes_dict.values())
+        edges = list(edges_dict.values())
+        # move rank to props, omit rank 0
+        for edge in edges:
+            if "rank" in edge:
+                rank = edge.pop("rank")
+                if rank != 0:
+                    edge["props"]["rank"] = rank
+
+        return {
+            "nodes": nodes,
+            "edges": edges,
+            "nodes_dict": nodes_dict,
+            "edges_dict": edges_dict,
+            "nodes_count": len(nodes),
+            "edges_count": len(edges),
+        }
+
+    def as_pandas(self, primitive: bool = True):
```

Comment:
it's better to use `as_data_frame`?
because maybe we can also use [Polars](https://pola.rs) which also supports `DataFrame`

Reply:
> it's better to use as_data_frame?

fair, df in Py world are mostly pd df, thus calling it as_data_frame() makes sense.

Wow, Polars is cool, but the pandas dependency here is not added due to we need to keep this repo slim.

Reply:
done